transformer_engine/pytorch/ops/fuser.py

```diff
@@ -60,39 +68,83 @@ def forward(
             Basic operations
         basic_op_kwargs: list of dict
             Keyword arguments to BasicOperation
-        *params: torch.nn.Parameter
-            Parameters in operation pipeline
+        num_params: int
+            Number of parameter tensors to include in autograd graph.
+        *extra_inputs: torch.Tensor
+            Extra tensor inputs to include in autograd graph. Consists
+            of parameter tensors, followed by extra operation inputs.
+
+        Returns
+        -------
+        Output tensor(s). If none of the operations have any extra
+        tensor outputs, then the pipeline's output tensor is returned.
+        Otherwise, a tuple with the pipeline's output tensor and extra
+        tensor outputs is returned.
 
         """
 
         # Operation autograd contexts
         basic_op_ctxs = [OperationContext() for _ in range(len(basic_ops))]
 
+        # Unflatten list of parameters and extra tensor inputs
+        num_extra_inputs = sum(op.num_extra_inputs for op in basic_ops)
```

Comment:
Why not calculate this once in the constructor instead?

Reply:
Constructor of sequential I mean. I think you do quite a lot of work in this autograd function that is not really changing from run to run, so we should move as much of this to the sequential constructor as possible.